include/commissioner/commissioner.hpp

```diff
@@ -127,35 +125,137 @@ struct Config
-    JoinerType mType;
+    /**
+     * This function notifies a joiner DTLS session is connected or not.
+     *
+     * @param[in] aJoinerId  The joiner ID.
+     * @param[in] aError     The error during connecting; the joiner connection
+     *                       is successfully established if error is Error::kNone;
+     *                       otherwise, it failed to connect.
+     *
+     */
+    virtual void OnJoinerConnected(const ByteArray &aJoinerId, Error aError)
+    {
+        (void)aJoinerId;
+        (void)aError;
+    }
+
+    /**
+     * This function notifies the receiving of JOIN_FIN.req message and asks for
+     * vendor-specific provisioning if required.
+     *
+     * @param[in]  aJoinerId           The joiner ID.
+     * @param[in]  aVendorName         A human-readable product vendor name string in utf-8 format.
+     * @param[in]  aVendorModel        A human-readable product model string.
+     * @param[in]  aVendorSwVersion    A utf-8 string that specifies the product software version.
+     * @param[in]  aVendorStackVersion A vendor stack version of fixed length (5 bytes). See section
+     *                                 8.10.3.6 of Thread spec for detail.
+     * @param[in]  aProvisioningUrl    A URL encoded as a utf-8 string provided by the Joiner
+     *                                 to communicate to the user which Commissioning application
+     *                                 is best suited to properly provision it to the appropriate
+     *                                 service. Empty if the joiner doesn't provide it.
+     * @param[in]  aVendorData         A product vendor-defined data structure to guide
+     *                                 vendor-specific provisioning. Empty if the joiner doesn't provide it.
+     *
+     * @return  A boolean indicates whether the joiner is accepted.
+     *
+     * @note This will be called when A well-formed JOIN_FIN.req has been received.
+     */
+    virtual bool OnJoinerFinalize(const ByteArray &  aJoinerId,
+                                  const std::string &aVendorName,
+                                  const std::string &aVendorModel,
+                                  const std::string &aVendorSwVersion,
+                                  const ByteArray &  aVendorStackVersion,
+                                  const std::string &aProvisioningUrl,
+                                  const ByteArray &  aVendorData)
+    {
+        (void)aJoinerId;
+        (void)aVendorName;
+        (void)aVendorModel;
+        (void)aVendorSwVersion;
+        (void)aVendorStackVersion;
+        (void)aProvisioningUrl;
+        (void)aVendorData;
+        return false;
+    }
 
-    // If the value is all-zeros, it represents for all joiners of this type.
-    uint64_t mEui64; ///< The IEEE EUI-64 value.
+    /**
+     * This funtions notifies the response of a keep-alive message.
+     *
+     * @param[in] aError  An error indicates whether the keep-alive message
```

Comment:
Clearly states what errors to expect?

Reply:
Giving the accurate errors is not easy. Please take a look at the call site of this handler https://github.com/openthread/ot-commissioner/pull/95/files#diff-664c9f59b79e0e70d3f1e8d022c9b49bR1151.

It propagates every error on the call chain, and we need to dig into every function called in the chain to aggregate a set of possible errors and this set will change when implementation changes.

Most of the errors in the commissioner interface serve indicating whether an operation has succeed or failed. This may not be the beast practice and we can discuss about it.

For this callback, we can normalize all possible errors to `Error::kNone` and `Error::kReject`, but it will drop the actual error, which can carry valuable debug information, during the keep-alive request-response exchange. Thoughts?

Reply:
Fine to not list all possible errors.